### PR TITLE
feat: base image config

### PR DESCRIPTION
**Pull request by yanksyoon**
Applicable spec: N/A

### Overview

Adds support for new LTS release (noble).

### Rationale

To support new noble image.
Please note that not all image supports are added to ensure that the charm works.

### Juju Events Changes

None.

### Module Changes

- `.github/workflows/integration_test.yaml` - updated syntax to avoid repetition.
- `scripts/build-lxd-image.sh`
  - modified to cater for base image support.
  - removed unnecessary 22.04 hwe kernel update.
- `scripts/build-openstack-image.sh` - modified to cater for base image support.
- `src/charm.py` - replace hardcoded jammy image to config.
- `src/charm_state.py`
  - fixed bug on github path parsing
  - `BaseImage` enum introduced w/ parsing
- `src/openstack_cloud/openstack_manager.py`
  - modified to cater for base images
  - a bit of refactoring
- tests
  - relevant tests
  - + a bit of tests sneaked in to increase coverage :D 

### Library Changes

None.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: docs/how-to/set-base-image.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# How to set base image
+
+This charm supports deploying the runners on different base images.
+
+By using [`juju config`](https://juju.is/docs/juju/juju-config) to change the
+[charm configuration labels](https://charmhub.io/github-runner/configure#base-image), the runner
+can be deployed with a different Ubuntu base image. The supported base images are limited to jammy
+and noble to ensure guaranteed capabilities.
```
Comment (jdkandersson) [inline docs/how-to/set-base-image.md:8]: Perhaps we just note that they are the most recent LTS images or something like that?

File: src-docs/charm_state.py.md [hunk 2]
```diff
@@ -333,6 +369,7 @@ Runner configurations for the charm.
 
 **Attributes:**
  
+ - <b>`base_image`</b>:  The ubuntu base image to run the runner viertual machines on.
```
Comment (jdkandersson) [inline src-docs/charm_state.py.md:372]: virtual?

File: src-docs/openstack_manager.md [hunk 3]
```diff
@@ -167,3 +166,32 @@ __init__(
...
 
 
+---
+
+<a href="../src/openstack_cloud/openstack_manager.py#L242"><img align="right" style="float:right;" src="https://img.shields.io/badge/-source-cccccc?style=flat-square"></a>
+
+## <kbd>class</kbd> `BuildImageConfig`
+The configuration values for building openstack image. 
+
+Attrs:  arch: The image architecture to build for.  base_image: The ubuntu image to use as image build base.  proxies: HTTP proxy settings.
```
Comment (jdkandersson) [inline src-docs/openstack_manager.md:176]: this should be `Attributes` so that it renders correctly 🙂

File: src/charm_state.py [hunk 4]
```diff
@@ -341,15 +342,52 @@ def check_fields(cls, values: dict) -> dict:
...
+        Raises:
+            ValueError if an unsupporte base image is passed in.
+
+        Returns:
+            The base image configuration of the charm.
+        """
+        image_name = charm.config.get(BASE_IMAGE_CONFIG_NAME, "jammy").lower().strip()
+        try:
+            return cls(image_name)
+        except ValueError as exc:
```
Comment (jdkandersson) [inline src/charm_state.py:374]: Perhaps we just do a get from `LTS_IMAGE_VERSION_TAG_MAP` and put the current value as the default?

File: src/charm_state.py [hunk 5]
```diff
@@ -341,15 +342,52 @@ def check_fields(cls, values: dict) -> dict:
...
+            if image_name in LTS_IMAGE_VERSION_TAG_MAP:
+                return cls(LTS_IMAGE_VERSION_TAG_MAP[image_name])
+            raise ValueError(f"Unsupported base image: {image_name}") from exc
+
+
 class RunnerCharmConfig(BaseModel):
     """Runner configurations for the charm.
 
     Attributes:
+        base_image: The ubuntu base image to run the runner viertual machines on.
```
Comment (jdkandersson) [inline src/charm_state.py:384]: virtual?

File: src/charm_state.py [hunk 6]
```diff
@@ -610,6 +666,36 @@ class CharmState:
...
+    ) -> None:
+        """Ensure immutable config has not changed.
+
+        Raises:
+            ImmutableConfigChangedError: If an immutable configuration has changed.
+        """
+        if not prev_state:
+            return
+        if prev_state["runner_config"]["runner_storage"] != runner_storage:
+            logger.warning(
```
Comment (jdkandersson) [inline src/charm_state.py:681]: should this be an error log?

File: tests/integration/test_charm_one_runner.py [hunk 7]
```diff
@@ -291,3 +296,40 @@ async def test_token_config_changed_insufficient_perms(
     await model.wait_for_idle()
 
     await wait_till_num_of_runners(unit, num=0)
+
+
+@pytest.mark.parametrize(
+    "image",
+    [
+        pytest.param("noble", id="noble"),
```
Comment (jdkandersson) [inline tests/integration/test_charm_one_runner.py:304]: should we leave this as not parametrised until we need o test for more ubuntu OS?

File: tests/integration/test_openstack.py [hunk 8]
```diff
@@ -55,3 +60,40 @@ async def test_openstack_integration(
...
     assert server.image.name == "jammy"
+
+
+@pytest.mark.parametrize(
+    "image",
+    [
+        pytest.param("noble", id="noble"),
+    ],
+)
+async def test_runner_base_image(
```
Comment (jdkandersson) [inline tests/integration/test_openstack.py:71]: similar comment as above

File: src/charm_state.py (hunk 6 above)
Comment (yanksyoon) [reply src/charm_state.py]: Changed to error log, thank you!

File: src/charm_state.py (hunk 5 above)
Comment (yanksyoon) [reply src/charm_state.py]: Fixed! thank you!

File: src/charm_state.py (hunk 4 above)
Comment (yanksyoon) [reply src/charm_state.py]: Fixed, thank you!

File: docs/how-to/set-base-image.md (hunk 1 above)
Comment (yanksyoon) [reply docs/how-to/set-base-image.md]: Updated, thank you!

File: src-docs/openstack_manager.md (hunk 3 above)
Comment (yanksyoon) [reply src-docs/openstack_manager.md]: Updated, thank you!

**Review by jdkandersson — DISMISSED** [review]

**Review by arturo-seijas — DISMISSED** [review]

**Review by arturo-seijas — DISMISSED** [review]

**Review by jdkandersson — DISMISSED** [review]

**Review by arturo-seijas — DISMISSED** [review]

**Review by jdkandersson — DISMISSED** [review]

**Review by jdkandersson — DISMISSED** [review]

File: docs/how-to/set-base-image.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
+# How to set base image
+
+This charm supports deploying the runners on different base images.
```
Comment (amandahla) [inline docs/how-to/set-base-image.md:3]: Minor: The default is jammy, right? Should be described here in this docs or is already described somewhere else?

Comment (yanksyoon) [reply docs/how-to/set-base-image.md:3]: That sounds like a good addition - the only place it mentions right now is in the `config.yaml`, thank you! (I'll push the change tomorrow morning - the runners seem to be very busy atm)

**Review by jdkandersson — DISMISSED** [review]

**Review by jdkandersson — APPROVED** [review]

**Review by merkata — APPROVED** [review]
Good stuff!

File: scripts/build-openstack-image.sh [hunk 10]
```diff
@@ -14,6 +14,7 @@ HTTPS_PROXY="$3"
 NO_PROXY="$4"
 DOCKER_PROXY_SERVICE_CONF="$5"
 DOCKER_PROXY_CONF="$6"
+BASE_IMAGE="$7"
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply scripts/build-openstack-image.sh:17]: Yup! This is re-written in Python in https://github.com/canonical/github-runner-image-builder and the image will be provided over the relation :D

**Pull request merged**